pandas/core/arrays/categorical.py

```diff
@@ -2441,19 +2441,31 @@ def replace(self, to_replace, value, inplace: bool = False):
         """
         inplace = validate_bool_kwarg(inplace, "inplace")
         cat = self if inplace else self.copy()
-        if to_replace in cat.categories:
-            if isna(value):
-                cat.remove_categories(to_replace, inplace=True)
-            else:
-                categories = cat.categories.tolist()
-                index = categories.index(to_replace)
-                if value in cat.categories:
-                    value_index = categories.index(value)
-                    cat._codes[cat._codes == index] = value_index
-                    cat.remove_categories(to_replace, inplace=True)
+
+        # build a dict of (to replace -> value) pairs
+        if is_list_like(to_replace):
+            # if to_replace is list-like and value is scalar
+            replace_dict = {replace_value: value for replace_value in to_replace}
+        else:
+            # if both to_replace and value are scalar
+            replace_dict = {to_replace: value}
+        # other cases, like if both to_replace and value are list-like or if
+        # to_replace is a dict, are handled separately in NDFrame
+
+        for replace_value, new_value in replace_dict.items():
+            if replace_value in cat.categories:
+                if isna(new_value):
```

Comment:
move the isna check first as it an never be true by definition (no nulls in categories); also simplifies the logic slightly

```

if isna(….):
    ….
   continue

….rest of logic
```

Reply:
@jreback This breaks a lot of other tests. I think isna should be allowed. The [documentation](https://pandas.pydata.org/pandas-docs/stable/user_guide/categorical.html) says "All values of categorical data are either in categories or np.nan"

PS I merged master

Reply:
you are missing my point. you simply need to reorganize, its the same condition w/o the dedent